plugins/woocommerce-blocks/tests/e2e/tests/product-collection/product-collection.block_theme.spec.ts

```diff
@@ -966,6 +966,63 @@ test.describe( 'Product Collection', () => {
 				.poll( async () => await page.evaluate( 'window.eventFired' ) )
 				.toBe( 3 );
 		} );
+
+		test.describe( 'wc-blocks_viewed_product is emitted', () => {
+			let promise: Promise< { productId?: number; collection?: string } >;
+
+			test.beforeEach( async ( { page, pageObject } ) => {
+				await pageObject.createNewPostAndInsertBlock( 'featured' );
+
+				promise = new Promise( ( resolve ) => {
+					void page.exposeFunction( 'resolvePayload', resolve );
+					void page.addInitScript( () => {
+						window.document.addEventListener(
+							'wc-blocks_viewed_product',
+							( e ) => {
+								window.resolvePayload( e.detail );
+							}
+						);
+					} );
+				} );
+
+				await pageObject.publishAndGoToFrontend();
+			} );
+
+			test( 'when Product Image is clicked', async ( { page } ) => {
+				await page
+					.locator( '[data-block-name="woocommerce/product-image"]' )
+					.nth( 0 )
+					.click();
+
+				const { collection, productId } = await promise;
+				expect( collection ).toEqual(
+					'woocommerce/product-collection/featured'
+				);
+				expect( productId ).toEqual( expect.any( Number ) );
+			} );
```

Comment:
nit: When an event is fired after an action is performed (click in this case), we don't have to bind via `addInitScript` – we can use `page.evaluate` in this case. For example:

```ts
const waitForViewedProductEvent = ( page ) => {
	return page.evaluate( () => {
		return new Promise( ( resolve ) => {
			window.addEventListener(
				'wc-blocks_viewed_product',
				resolve
			);
		} );
	} );
};

test( 'when Product Image is clicked', async ( { page } ) => {
	const eventPromise = waitForViewedProductEvent( page );

	await page
		.getByLabel( 'Block: Product Image' )
		.first()
		.click();

	const event = await eventPromise;

	expect( event.collection ).toEqual(
		'woocommerce/product-collection/featured'
	);
	expect( event.productId ).toEqual( expect.any( Number ) );
} );
```

This way it resembles more how we use the Playwright's `page.waitForRequest` and similar APIs. Also, if used extensively, `page.waitForEvent` could be exposed as a part of `PageUtils` fixture.

Reply:
I think I can't do this due to redirect that happens on click. So I'm losing a point of reference with the `page.evaluate` 🙌